### ISSUE TITLE
Ubuntu kontena-agent package is not compatible with Docker 17.06

### ISSUE THREAD

**Issue by roquie** [kontena/kontena#2588]
### Env
OS: Ubuntu 16.04 LTS
Docker: 17.x

### Steps to reproduce:
1. Install docker 17.x
2. After step by step follow documentation https://kontena.io/docs/getting-started/installing/ubuntu.html

When `kontena-agent` installation on server is come on to end, post script run a `service docker restart` and then docker is fall with error. Thats all.

My guess-works:
Inside a `/etc/systemd/system/docker.service.d/50-kontena.conf` run of docker in this manner deprecated and was removed:
```
//...
ExecStart=/usr/bin/docker daemon -H fd:// ${DOCKER_OPTS}
```

This command returned an error: 
```
"docker daemon" is not supported on Linux. Please run "dockerd" directly
```
So, process it was crashed.

With `docker.io=12.*` agent perfectly works.

**Comment by SpComb** [comment]
The `ExecStart=/usr/bin/docker daemon` was originally written for compatibility with Docker 1.8+, in order to provide the `${DOCKER_OPTS}` for injecting the `--insecure-registry="10.81.0.0/16"` needed for the kontena registry to work.

It looks like support for the older `docker daemon` CLI command was removed in 17.06: https://github.com/docker/for-linux/issues/11 https://github.com/moby/moby/issues/33931 ([deprecated](https://docs.docker.com/engine/deprecated/#docker-daemon-subcommand) in 1.13, scheduled for removal in... 17.12?). I think older 17.x releases should still work?

Clearly time to revisit the `docker.service `dropin that the Ubuntu package uses, to either use `ExecStart=/usr/bin/dockerd ... ${DOCKER_OPTS}`, or see if we can use something else to configure the registry. We have since bumped the minimum Docker version to 1.10, but still need to figure out which versions of which Docker package variants include support for the new `dockerd` binary.

**Comment by SpComb** [comment]
> We have since bumped the minimum Docker version to 1.10, but still need to figure out which versions of which Docker package variants include support for the new dockerd binary.

It looks like the Ubuntu xenial `docker.io` `1.10.3-0ubuntu6` package does not include `/usr/bin/dockerd`. All the other package variants that I use for testing support both `/usr/bin/docker daemon` as well as `/usr/bin/dockerd`:

* Ubuntu xenial `docker.io` `1.12.6-0ubuntu1~16.04.1`
* `https://apt.dockerproject.org/repo ubuntu-xenial/main` `docker-engine` `1.13.1-0~ubuntu-xenial`
* `https://download.docker.com/linux/ubuntu xenial/stable` `docker-ce` `17.03.2~ce-0~ubuntu-xenial`

The `docker.io` package uses `ExecStart=/usr/bin/dockerd -H fd:// $DOCKER_OPTS`, whereas the `docker-engine` and `docker-ce` packages use `ExecStart=/usr/bin/dockerd -H fd://`.

Upgrading to the latest available versions:

* `docker-engine` `17.05.0~ce-0~ubuntu-xenial` still works with `/usr/bin/docker daemon`
* `docker-ce` `17.06.0~ce-0~ubuntu` fails with `/usr/bin/docker daemon`

Both packages still use `/usr/bin/dockerd -H fd://`.

**Comment by SpComb** [comment]
Looking at the older available versions:

* `docker.io` `1.10.3-0ubuntu6` uses `/usr/bin/docker daemon` and does NOT support `/usr/bin/dockerd`
* `docker-engine` `1.10.*` is no longer available
* `docker-engine` `1.11.2-0~xenial` uses `/usr/bin/docker daemon` and does NOT support `/usr/bin/dockerd`
* `docker-engine` `1.12.6-0~ubuntu-xenial` uses `/usr/bin/dockerd` and still supports `/usr/bin/docker daemon`

So I reckon we drop support for Docker 1.10, bump the minimum Docker version to 1.12, and switch to using `ExecStart=/usr/bin/dockerd`, which is compatible with the newer Docker 17.06 version.

@jakolehm WDYT?

**Comment by jakolehm** [comment]
We cannot bump minimum Docker version in a patch release.. seems that we need to move milestone to 1.4.0?

**Comment by SpComb** [comment]
One approach might be the #2589 minimum version bump to use `ExecStart=/usr/bin/dockerd` for 1.4, and then add a maximum version constraint on `docker-ce << 17.06` for 1.3.4... not sure how easy that is with the Debian `Depends:` syntax, need to try.

**Comment by roquie** [comment]
@SpComb Nice study of task, I like it :)

**Comment by roquie** [comment]
I hurried and did not correctly understand the fix. Issue should be still open, because 1.3.4 version just set borders (>=1.10 & <17.06) of docker version. This issue will be closed after 1.4.0, when docker 17.06 will work with kontena-agent.

@jakolehm Everything is correct? :)

**Comment by SpComb** [comment]
> I hurried and did not correctly understand the fix. Issue should be still open, because 1.3.4 version just set borders (>=1.10 & <17.06) of docker version. This issue will be closed after 1.4.0, when docker 17.06 will work with kontena-agent.

Yes, I'll close this once #2589 is merged for 1.4.0.dev

**Comment by roquie** [comment]
> Yes, I'll close this once #2589 is merged for 1.4.0.dev

Okay
